tenseal/cpp/tensors/ckksvector.cpp

```diff
@@ -564,39 +564,47 @@ CKKSVector& CKKSVector::polyval_inplace(const vector<double>& coefficients) {
     return *this;
 }
 
-CKKSVector CKKSVector::conv2d_im2col(const vector<double>& kernel,
-                                     size_t windows_nb) {
+CKKSVector CKKSVector::conv2d_im2col(const vector<vector<double>>& kernel,
+                                     const size_t windows_nb) {
     CKKSVector new_vec = *this;
     new_vec.conv2d_im2col_inplace(kernel, windows_nb);
     return new_vec;
 }
 
-CKKSVector& CKKSVector::conv2d_im2col_inplace(const vector<double>& kernel,
-                                              const size_t windows_nb) {
-    vector<double> plain_vec;
-    size_t chunks_nb = kernel.size();
-
+CKKSVector& CKKSVector::conv2d_im2col_inplace(
+    const vector<vector<double>>& kernel, const size_t windows_nb) {
     if (windows_nb == 0) {
         throw invalid_argument("Windows number can't be zero");
     }
 
-    if (kernel.empty()) {
-        throw invalid_argument("Kernel vector can't be empty");
+    if (kernel.empty() ||
+        (any_of(kernel.begin(), kernel.end(),
+                [](const vector<double>& i) { return i.empty(); }))) {
+        throw invalid_argument("Kernel matrix can't be empty");
```

Comment:
What about something like this? to check whether the vector<vector> is a well organized matrix?

```suggestion
    if (kernel.empty() ||
        (any_of(kernel.begin(), kernel.end(),
                [](const vector<double>& i) { return i.size() == kernel[0].size(); }))) {
        throw invalid_argument("Kernel matrix can't be empty");
```

Reply:
The test will be wrong in the case where `kernel[0].size() == 0`.
in other hand `horizontal_scan` always check if the rows have the same size.